be/src/vec/exec/join/vjoin_node_base.cpp

```diff
@@ -183,16 +192,21 @@ Status VJoinNodeBase::_build_output_block(Block* origin_block, Block* output_blo
             for (int i = 0; i < mutable_columns.size(); ++i) {
                 auto result_column_id = -1;
                 RETURN_IF_ERROR(_output_expr_ctxs[i]->execute(origin_block, &result_column_id));
-                auto column_ptr = origin_block->get_by_position(result_column_id)
-                                          .column->convert_to_full_column_if_const();
-                insert_column_datas(mutable_columns[i], *column_ptr, rows);
+                auto& origin_column = origin_block->get_by_position(result_column_id).column;
+                auto column_ptr = origin_column->convert_to_full_column_if_const();
+                if (column_ptr.get() == origin_column.get()) {
+                    column_ptr = nullptr;
+                    insert_column_datas(mutable_columns[i], origin_column, rows);
+                } else {
+                    insert_column_datas(mutable_columns[i], column_ptr, rows);
+                }
             }
         }
 
-        if (!is_mem_reuse) {
+        if (!is_mem_reuse || !keep_origin) {
             output_block->swap(mutable_block.to_block());
         }
-        DCHECK(output_block->rows() == rows);
```

Comment:
`DCHECK` is more reasonable. If you want an error reported to users, you could return a error Status here.